crates/core/tedge_monitor/src/main.rs

```diff
@@ -0,0 +1,166 @@
+            }));
+        } else if service.contains("az") {
+            watchdog_threads.push(thread::spawn(move || {
+                monitor_tedge_service("tedge_mapper", Some("az"), &req_topic, &res_topic)
+            }));
+        } else if service.contains("collectd") {
+            watchdog_threads.push(thread::spawn(move || {
+                monitor_tedge_service("tedge_mapper", Some("collectd"), &req_topic, &res_topic)
+            }));
+        } else {
+            watchdog_threads.push(thread::spawn(move || {
+                monitor_tedge_service("tedge_agent", None, &req_topic, &res_topic)
+            }));
+        }
+    }
+    for child in watchdog_threads {
+        // Wait for the thread to finish. Returns a result.
+        let _ = child.join();
+    }
+}
+
+pub fn monitor_tedge_service(
+    name: &str,
+    m_type: Option<&str>,
+    req_topic: &str,
+    res_topic: &str,
+) -> Result<ExitStatus, WatchdogError> {
+    let client_id: &str = &format!("{}-{:?}-health", name, m_type.ok_or("None"));
+    let options = get_mqtt_options(client_id);
+    dbg!(&options);
+    let (mut client, mut connection) = rumqttc::Client::new(options, 10);
+
+    client.subscribe(res_topic, AtLeastOnce).unwrap();
+
+    loop {
+        if get_process_id(name, m_type) > 0 {
+            for event in connection.iter() {
+                match event {
+                    Ok(Event::Incoming(Packet::SubAck(_))) => {
+                        client.publish(req_topic, AtLeastOnce, false, "").unwrap();
+                    }
+                    Ok(Event::Incoming(Packet::PubAck(_))) => {
+                        continue;
+                        // The request has been sent
+                    }
+                    Ok(Event::Incoming(Packet::Publish(_response))) => {
+                        // Received response from thin-edge service, update the status to systemd on behalf of service
+                        let pid = get_process_id(name, m_type);
+                        notify_systemd(pid, "WATCHDOG=1")?;
+                        break;
+                    }
+                    Ok(Event::Outgoing(Outgoing::PingReq)) => {
+                        // No messages have been received for a while
+                        eprintln!("Local MQTT publish has timed out.");
+                        break;
+                    }
+                    Ok(Event::Incoming(Incoming::Disconnect)) => {
+                        eprintln!("ERROR: Disconnected");
+                        break;
+                    }
+                    Err(err) => {
+                        eprintln!("ERROR: {:?}", err);
+                        // client.subscribe(res_topic, AtLeastOnce).unwrap();
+                        break;
+                    }
+                    _ => {}
+                }
+            }
+        } else {
+            thread::sleep(time::Duration::from_secs(5));
+        }
+        thread::sleep(time::Duration::from_secs(2));
```

Comment:
These threads sleep a lot, maybe this could be made async?

Reply:
Sure.